File: packages/protocol/python/person_protocol/validator.py
"""Validate protocol messages against the canonical JSON Schema files.

The Node runtime compiles the same files. Keeping one schema set and two thin
bindings is what stops the TypeScript and Python views of the contract from
drifting apart.
"""

from __future__ import annotations

import json
from collections.abc import Iterable
from functools import lru_cache
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator
from jsonschema.exceptions import ValidationError
from referencing import Registry, Resource

from .assets import package_asset_directory
from .version import MESSAGE_TYPES, PROTOCOL_VERSION, SCHEMA_FILES


class ProtocolError(ValueError):
    """A message did not satisfy the versioned protocol contract."""

    def __init__(self, message: str, diagnostics: Iterable[str] = ()) -> None:
        self.diagnostics = list(diagnostics)
        detail = "; ".join(self.diagnostics)
        super().__init__(f"{message}: {detail}" if detail else message)


def schema_directory() -> Path:
    return package_asset_directory(__file__, "schemas")


def _describe(error: ValidationError) -> str:
    location = "/" + "/".join(str(part) for part in error.absolute_path)
    return f"{location} {error.message}"


class ProtocolValidator:
    def __init__(self, directory: Path | None = None) -> None:
        self.directory = directory or schema_directory()
        resources = []
        for path in sorted(self.directory.glob("*.schema.json")):
            schema = json.loads(path.read_text(encoding="utf-8"))
            resources.append((path.name, Resource.from_contents(schema)))
        registry = Registry().with_resources(resources)
        self._validators: dict[str, Draft202012Validator] = {}
        for message_type in MESSAGE_TYPES:
            filename = SCHEMA_FILES[message_type]
            schema = json.loads((self.directory / filename).read_text(encoding="utf-8"))
            self._validators[message_type] = Draft202012Validator(
                schema,
                registry=registry,
                format_checker=Draft202012Validator.FORMAT_CHECKER,
            )

    @property
    def message_types(self) -> tuple[str, ...]:
        return MESSAGE_TYPES

    def validate(self, message: Any) -> tuple[bool, list[str]]:
        if not isinstance(message, dict):
            return False, ["/ message must be a JSON object"]
        version = message.get("protocolVersion")
        if version != PROTOCOL_VERSION:
            return False, [
                f"/protocolVersion unsupported protocol version {json.dumps(version)}; "
                f"this runtime speaks {PROTOCOL_VERSION}"
            ]
        message_type = message.get("type")
        if not isinstance(message_type, str) or message_type not in self._validators:
            return False, [f"/type unknown message type {json.dumps(message_type)}"]
        errors = sorted(
            self._validators[message_type].iter_errors(message),
            key=lambda error: list(error.absolute_path),
        )
        if errors:
            return False, [_describe(error) for error in errors]
        return True, []

    def assert_valid(self, message: Any) -> dict[str, Any]:
        valid, diagnostics = self.validate(message)
        if not valid:
            raise ProtocolError("Invalid protocol message", diagnostics)
        assert isinstance(message, dict)
        return message


@lru_cache(maxsize=1)
def protocol_validator() -> ProtocolValidator:
    """Process-wide validator; compiling the schema set once is much cheaper."""
    return ProtocolValidator()

File: packages/protocol/ts/validator.ts
import { readFileSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";
import { Ajv2020 } from "ajv/dist/2020.js";
import type { ErrorObject, ValidateFunction } from "ajv";
import addFormatsModule from "ajv-formats";

// ajv-formats is CommonJS: its module.exports is the plugin function itself,
// which NodeNext hands back as the default import at runtime.
const addFormats = addFormatsModule as unknown as (ajv: Ajv2020) => unknown;
import {
  MESSAGE_TYPES,
  PROTOCOL_VERSION,
  SCHEMA_FILES,
  type MessageType,
} from "./version.ts";
import type { ProtocolMessage } from "./types.ts";

export const SCHEMA_DIRECTORY = fileURLToPath(
  new URL("../schemas/", import.meta.url),
);

export class ProtocolError extends Error {
  readonly diagnostics: string[];
  constructor(message: string, diagnostics: string[] = []) {
    super(
      diagnostics.length ? `${message}: ${diagnostics.join("; ")}` : message,
    );
    this.name = "ProtocolError";
    this.diagnostics = diagnostics;
  }
}

export interface ValidationResult {
  valid: boolean;
  diagnostics: string[];
}

const describe = (errors: ErrorObject[] | null | undefined): string[] =>
  (errors ?? []).map(
    (error) => `${error.instancePath || "/"} ${error.message ?? "is invalid"}`,
  );

/**
 * Validates protocol messages against the canonical JSON Schema files that the
 * Python runtime also loads. Both runtimes therefore enforce one contract
 * rather than two hand-maintained approximations of it.
 */
export class ProtocolValidator {
  readonly #ajv: Ajv2020;
  readonly #byType = new Map<MessageType, ValidateFunction>();

  constructor(schemaDirectory: string = SCHEMA_DIRECTORY) {
    this.#ajv = new Ajv2020({
      strict: true,
      allErrors: true,
      allowUnionTypes: true,
    });
    addFormats(this.#ajv);
    for (const file of readdirSync(schemaDirectory).sort()) {
      if (!file.endsWith(".schema.json")) continue;
      const schema = JSON.parse(
        readFileSync(path.join(schemaDirectory, file), "utf8"),
      ) as object;
      this.#ajv.addSchema(schema, file);
    }
    for (const type of MESSAGE_TYPES) {
      const compiled = this.#ajv.getSchema(SCHEMA_FILES[type]);
      if (!compiled)
        throw new ProtocolError(`Missing schema for message type ${type}`);
      this.#byType.set(type, compiled);
    }
  }

  /** Message types this validator knows about, in a stable order. */
  get messageTypes(): readonly MessageType[] {
    return MESSAGE_TYPES;
  }

  validate(message: unknown): ValidationResult {
    if (
      typeof message !== "object" ||
      message === null ||
      Array.isArray(message)
    )
      return { valid: false, diagnostics: ["/ message must be a JSON object"] };
    const record = message as Record<string, unknown>;
    if (record["protocolVersion"] !== PROTOCOL_VERSION)
      return {
        valid: false,
        diagnostics: [
          `/protocolVersion unsupported protocol version ${JSON.stringify(
            record["protocolVersion"],
          )}; this runtime speaks ${PROTOCOL_VERSION}`,
        ],
      };
    const type = record["type"];
    if (typeof type !== "string" || !this.#byType.has(type as MessageType))
      return {
        valid: false,
        diagnostics: [`/type unknown message type ${JSON.stringify(type)}`],
      };
    const check = this.#byType.get(type as MessageType) as ValidateFunction;
    return check(message)
      ? { valid: true, diagnostics: [] }
      : { valid: false, diagnostics: describe(check.errors) };
  }

  assertValid<T extends ProtocolMessage>(message: unknown): T {
    const result = this.validate(message);
    if (!result.valid)
      throw new ProtocolError("Invalid protocol message", result.diagnostics);
    return message as T;
  }
}

let shared: ProtocolValidator | undefined;

/** Process-wide validator. Compiling the schema set once is measurably cheaper. */
export function protocolValidator(): ProtocolValidator {
  shared ??= new ProtocolValidator();
  return shared;
}
